MixinServices/MixinServices/Database/User/DAO/MessageDAO.swift

```diff
@@ -980,18 +980,96 @@ extension MessageDAO {
                                silentNotification: silentNotification)
     }
     
-    public func messages(limit: Int, after messageId: String?) -> [Message] {
-        var sql = "SELECT * FROM messages"
-        if let messageId {
-            sql += " WHERE ROWID > IFNULL((SELECT ROWID FROM messages WHERE id = '\(messageId)'), 0)"
+    public func messages(limit: Int, after rowID: Int, matching conversationIDs: [String]?) -> [Message] {
+        var sql = "SELECT * FROM messages WHERE ROWID > ?"
+        if let conversationIDs {
+            let ids = conversationIDs.joined(separator: "', '")
+            sql += " AND conversation_id IN ('\(ids)')"
         }
         sql += " ORDER BY ROWID LIMIT ?"
-        return db.select(with: sql, arguments: [limit])
+        return db.select(with: sql, arguments: [rowID, limit])
     }
     
-    public func messagesCount() -> Int {
-        let count: Int? = db.select(with: "SELECT COUNT(*) FROM messages")
-        return count ?? 0
+    public func messageRowID(createdAt: String) -> Int? {
+        db.select(with: "SELECT ROWID FROM messages WHERE created_at >= ? ORDER BY ROWID LIMIT 1", arguments: [createdAt])
```

Comment:
It is better to explicitly specify the ASC sorting.

Reply:
Update with [398a9da](https://github.com/MixinNetwork/ios-app/pull/1303/commits/398a9da8918bad3de6aa385fd86b70e9611edd66)